[JSC] Implement String#isWellFormed and String#toWellFormed

Constellation opened the pull request:
#### 6ccd7405a513dab05d5a59359330af76e177f968
<pre>
[JSC] Implement String#isWellFormed and String#toWellFormed
<a href="https://bugs.webkit.org/show_bug.cgi?id=248588">https://bugs.webkit.org/show_bug.cgi?id=248588</a>
rdar://problem/102849210

Reviewed by Ross Kirsling.

This patch implements String#isWellFormed and String#toWellFormed proposal[1].
String#isWellFormed returns true if String does not have wrong surrogate pairs (e.g. non-paired surrogate).
String#toWellFormed returns a string, which replaces non well-formed characters with unicode replacement character (\ufffd).

[1]: <a href="https://github.com/tc39/proposal-is-usv-string">https://github.com/tc39/proposal-is-usv-string</a>

* JSTests/stress/string-well-formed.js: Added.
(shouldBe):
* LayoutTests/js/Object-getOwnPropertyNames-expected.txt:
* LayoutTests/js/script-tests/Object-getOwnPropertyNames.js:
* Source/JavaScriptCore/runtime/CommonIdentifiers.h:
* Source/JavaScriptCore/runtime/OptionsList.h:
* Source/JavaScriptCore/runtime/StringPrototype.cpp:
(JSC::StringPrototype::finishCreation):
(JSC::nonWellFormedIndex):
(JSC::JSC_DEFINE_HOST_FUNCTION):

Canonical link: <a href="https://commits.webkit.org/257250@main">https://commits.webkit.org/257250@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/e76618e173768656702bfc3d85a6d4cc6544b587

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/98346 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/7553 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/31480 "Built successfully") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/8/builds/107791 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 🧪 win~~](https://ews-build.webkit.org/#/builders/10/builds/168063 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/8060 "Built successfully") | [  ~~mac-debug~~](https://ews-build.webkit.org/#/builders/71/builds/36309 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 gtk~~](https://ews-build.webkit.org/#/builders/36/builds/90915 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/12/builds/104434 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/103999 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/6127 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/84934 "Built successfully") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/90915 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/87943 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/3/builds/89688 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/36/builds/90915 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [  ~~🛠 🧪 jsc~~](https://ews-build.webkit.org/#/builders/1/builds/89168 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/81/builds/1506 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/73/builds/21052 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/38/builds/84911 "Built successfully") | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/82/builds/1443 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/70/builds/22575 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/46/builds/28665 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/6346 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/60/builds/45064 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/37/builds/87757 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/75/builds/2493 "Built successfully and passed tests") | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/79/builds/2801 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/62/builds/41988 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 jsc-mips-tests~~](https://ews-build.webkit.org/#/builders/45/builds/19667 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
<!--EWS-Status-Bubble-End-->

rkirsling commented on Source/JavaScriptCore/runtime/OptionsList.h:
```diff
@@ -553,6 +553,7 @@ bool canUseWebAssemblyFastMemory();
     v(Bool, useResizableArrayBuffer, true, Normal, "Expose ResizableArrayBuffer feature.") \
     v(Bool, useSharedArrayBuffer, false, Normal, nullptr) \
     v(Bool, useShadowRealm, false, Normal, "Expose the ShadowRealm object.") \
+    v(Bool, useStringWellFormed, true, Normal, "Expose the ShadowRealm object.") \
```
Oops, looks like the description was copy-pasted. :wink:

Constellation replied:
Fixed.

rkirsling commented on Source/JavaScriptCore/runtime/StringPrototype.cpp:
```diff
@@ -1757,4 +1764,115 @@ JSC_DEFINE_HOST_FUNCTION(stringProtoFuncNormalize, (JSGlobalObject* globalObject
+    RETURN_IF_EXCEPTION(scope, { });
+
+    if (string.is8Bit())
+        return JSValue::encode(jsBoolean(true)); // Latin-1 characters do not have surrogates.
```
Since this is the second 8-bit check in the function, maybe the comment should occur earlier?

Constellation replied:
Done.

rkirsling commented on Source/JavaScriptCore/runtime/StringPrototype.cpp:
```diff
@@ -1757,4 +1764,115 @@ JSC_DEFINE_HOST_FUNCTION(stringProtoFuncNormalize, (JSGlobalObject* globalObject
+
+    const UChar* characters = string.characters16();
+    unsigned length = string.length();
+    auto firstNonWellFormedIndex = nonWellFormedIndex(characters, length);
```
I think we could use the word "ill-formed" for the opposite of "well-formed", since that's what the proposal does.

Constellation replied:
Nice, fixed.

rkirsling commented on Source/JavaScriptCore/runtime/StringPrototype.cpp:
```diff
@@ -1757,4 +1764,115 @@ JSC_DEFINE_HOST_FUNCTION(stringProtoFuncNormalize, (JSGlobalObject* globalObject
+        }
+
+        if (U16_IS_SURROGATE_TRAIL(character)) {
+            buffer.append(replacementCharacter);
```
Wow, it really surprised me that `replacementCharacter` is the name of a global. 😅 (Since it doesn't really sound like a _specific_ character...)

Constellation committed:
[JSC] Implement String#isWellFormed and String#toWellFormed
https://bugs.webkit.org/show_bug.cgi?id=248588
rdar://problem/102849210

Reviewed by Ross Kirsling.

This patch implements String#isWellFormed and String#toWellFormed proposal[1].
String#isWellFormed returns true if String does not have wrong surrogate pairs (e.g. non-paired surrogate).
String#toWellFormed returns a string, which replaces non well-formed characters with unicode replacement character (\ufffd).

[1]: https://github.com/tc39/proposal-is-usv-string

* JSTests/stress/string-well-formed.js: Added.
(shouldBe):
* LayoutTests/js/Object-getOwnPropertyNames-expected.txt:
* LayoutTests/js/script-tests/Object-getOwnPropertyNames.js:
* Source/JavaScriptCore/runtime/CommonIdentifiers.h:
* Source/JavaScriptCore/runtime/OptionsList.h:
* Source/JavaScriptCore/runtime/StringPrototype.cpp:
(JSC::StringPrototype::finishCreation):
(JSC::nonWellFormedIndex):
(JSC::JSC_DEFINE_HOST_FUNCTION):

Canonical link: https://commits.webkit.org/257250@main
